### ISSUE TITLE
external markdown - data-line-numbers

### ISSUE THREAD

**Issue by gamgamstyle** [hakimel/reveal.js#2371]
Is it possible to show line numbers in the code section of external markdown files (data-markdown)?

**Comment by DatStorm** [comment]
Would love this feature currently I am trying 

```
``` javascript <!-- {_data-highlight-lines="2-3"} -->
console.log("test");
let a = 10;
const b = a + 10;

``` 
```

**Comment by oheydrew** [comment]
I came here searching for the ability to highlight/brighten specific lines of code within a code-block step-by-step, something currently available in the HTML revealjs implementation, but seemingly not in the markdown syntax.

Regular HTML Syntax works just fine:
```
<pre><code class="hljs" data-line-numbers="4|5">
import React, { useState } from 'react';

export const Example = () => (
  <div>This line is highlighted (brightened) first</div>
  <div>This line is highlighted (brightened) second</div>
  <div>This (and all other) lines are dimmed</div>
)
</code></pre>
```

Markdown Syntax, however - or something approximating, does <b>not</b> work:
```
```javascript <!-- .element: data-line-numbers="4|5" -->
import React, { useState } from 'react';

export const Example = () => (
  <div>This line is highlighted (brightened) first</div>
  <div>This line is highlighted (brightened) second</div>
  <div>This (and all other) lines are dimmed</div>
)
```.
```

Attempting to ignore the ` ``` ` markdown syntax & use HTML tags in markdown does not work,
either. 

I believe the only way to do this at the moment is to stick with the html/css revealJS format - and use the `<pre><code>` syntax.

**Comment by hakimel** [comment]
We should definitely support code line numbers in markdown but after some quick poking around I'm not sure how we can best implement it. Would be great if someone wants to attempt to tackle this.

**Comment by mircodezorzi** [comment]
After a quick look at the code base I managed to put together the following proof-of-concept patch:

``` diff
diff --git a/plugin/markdown/markdown.js b/plugin/markdown/markdown.js
index c641d81..5f2732a 100755
--- a/plugin/markdown/markdown.js
+++ b/plugin/markdown/markdown.js
@@ -412,10 +412,17 @@
 				throw 'The reveal.js Markdown plugin requires marked to be loaded';
 			}
 
+			const renderer = {
+				code(code, infostring, escaped) {
+					return `<pre><code data-line-numbers="4,8-11">${code}</code></pre>`;
+				}
+			};
+
 			if( typeof hljs !== 'undefined' ) {
 				marked.setOptions({
+					renderer: renderer,
 					highlight: function( code, lang ) {
-						return hljs.highlightAuto( code, [lang] ).value;
+						return code;
 					}
 				});
 			}
@@ -426,7 +433,7 @@
 				marked.setOptions( options );
 			}
 
-			return processSlides().then( convertSlides );
+			return processSlides().then( convertSlides ).then( RevealHighlight.init );
 
 		},
```

Unfortunately I'm not well versed in JavaScript so I'm not sure how to take it from here. The main problem with this patch is that `RevealHighlight.init` get's called twice and any code block declared in HTML will be corrupted.

If anyone has some suggestions I would be really happy to help to get this issue closed!

**Comment by hakimel** [comment]
That’s really helpful, @mircodezorzi! I’ll try it out and see how this can be built into the plugin soon and let you know how it goes.

**Comment by mircodezorzi** [comment]
If it's any help I managed to fix the re-highlighting issue although this creates a new bug:
```diff
@@ -384,6 +386,12 @@
              section.parentNode.getAttribute( 'data-attributes' ) ||
              DEFAULT_SLIDE_ATTRIBUTES_SEPARATOR);
 
+     // Highlight all code blocks
+     [].slice.call( section.querySelectorAll( '.reveal pre code' ) ).forEach( function( block ) {
+       RevealHighlight.highlightBlock( block );
+     });
+
+
      // If there were notes, we need to re-add them after
      // having overwritten the section's HTML
      if( notes ) {
```
Unfortunately the following snippets doesn't properly get escaped:
```html
<section data-markdown>
  <textarea data-template>
    ```javascript
import React, { useState } from 'react';

function Example() {
  const [count, setCount] = useState(0);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button onClick={() => setCount(count + 1)}>
        Click me
      </button>
    </div>
  );
}
    ```
  </textarea>
</section>

```

**Comment by mircodezorzi** [comment]
I think I managed to come up with a decent solution, I pushed a commit to my local fork [here](https://github.com/mircodezorzi/reveal.js/commit/b1e059e8e03c2de876c24af731c8ce44d8ff802b).

Should I open a pull request?

**Comment by hakimel** [comment]
@mircodezorzi Thanks but no need to open a PR. I've made significant changes to the markdown plugin in 4.0 (see the dev branch) so the code you're updating is outdated. Going to start working on this now and will let you know how it goes.

**Comment by hakimel** [comment]
Support for line numbers and highlights has been added (951f5d04c2324093c0f138ff2945acc58c6d3fd2) and will be available in 4.0 🥳

This will first highlight lines 1 and 4, and then on a separate step highlight lines 4 through 8:
```markdown
```javascript [1,4|4-8]
```

If you want to enable line numbers without highlighting a specific line use:
```markdown
```javascript []
```

I'm in the middle of moving all docs to revealjs.com so documentation for his will follow later. Let me know if anyone has any input on the syntax.
